ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,612 @@
+ * of that packet.
+ */
+static int ssl_queue_data(QUIC_STREAM *stream,
+                          const unsigned char *data, uint64_t data_len)
+{
+    return 1;
+}
+
+/*
+ * ASSUMPTION: ssl_close_stream() detaches the QUIC_STREAM from the
+ * QUIC_CONNECTION it's attached to, and then destroys that QUIC_STREAM
+ * (as well as its SSL object).
+ */
+static int ssl_close_stream(QUIC_STREAM *stream)
+{
+    return 1;
+}
+
+/*
+ * ASSUMPTION: ssl_close_connection() closes all the streams that are
+ * attached to it, then closes the QUIC_CONNECTION as well.
+ * Actual cleanup / destruction of the QUIC_CONNECTION is assumed to be done
+ * higher up in the call stack (state machine, for example?).
+ */
+static int ssl_close_connection(QUIC_CONNECTION *connection)
+{
+    return 1;
+}
+
+/*
+ * ASSUMPTION: ossl_statm_set_error_state() sets an overall error state in
+ * the state machine.  It's up to the state machine to determine what to do
+ * with it.
+ */
+#define QUIC_STREAM_STATE_ERROR 1
+static int ossl_statm_set_error_state(OSSL_STATM *statm, int errcode)
+{
+    return 1;
+}
+
+/* =====END TO BE REMOVED===== */
+
+enum depack_modes {
+    collect_ack_data,
+    process_frames
+};
+
+static int depack_process_frames(QUIC_CONNECTION *connection, PACKET *pkt,
+                                 int packet_space, OSSL_TIME received,
+                                 enum depack_modes mode,
+                                 OSSL_ACKM_RX_PKT *ackm_data)
+{
+    while (PACKET_remaining(pkt) > 0) {
+        uint64_t frame_type;
+
+        if (!ossl_quic_wire_peek_frame_header(pkt, &frame_type))
+            return 0;
+
+        switch (frame_type) {
+        case OSSL_QUIC_FRAME_TYPE_PING:
+            if (mode == collect_ack_data)
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+            /* FALLTHRU */
+        case OSSL_QUIC_FRAME_TYPE_PADDING:
+            /*
+             * We ignore this.
+             * For the PING frame, eliciting an ACK is enough.
+             * Since the PING frame is only a header, we simply
+             * skip it.
+             */
+            ossl_quic_wire_skip_frame_header(pkt, &frame_type);
+            break;
+
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITH_ECN: {
+            OSSL_QUIC_FRAME_ACK ack;
+            OSSL_QUIC_ACK_RANGE *ack_ranges;
+            uint64_t total_ranges = 0;
+            int ok = 1;          /* Assume the best */
+
+            if (!ossl_quic_wire_peek_frame_ack_num_ranges(pkt, &total_ranges)
+                || (ack_ranges = OPENSSL_zalloc(sizeof(ack_ranges[0])
+                                                * total_ranges)))
+                return 0;
+
+            ack.ack_ranges = ack_ranges;
+            ack.num_ack_ranges = total_ranges;
+
+            /* QUESTION: where do I get the ack delay exponent? */
```

Comment:
This is a question for @hlandau I guess.

Reply:
The ACK Delay Exponent is negotiated via a transport parameter (note that each direction can have a different value). The best solution here would probably be to provide a function to let whatever instantiated the RX depacketiser set the value. For testing, 3 is a commonly used value.

Reply:
I assume that there would also be a function with which I can *get* that value...  that's the interesting part here.

Reply:
If it's negotiated as a transport parameter then presumably you can get it from the QUIC_CONNECTION object somehow (`QUIC_CONNECTION_get_ack_delay_exponent()`). Although what do we use prior to the point that the parameter negotiation is complete? I guess there is a default value that such a function would return if it hasn't been set yet.

Reply:
That sounds like an idea.  I can create a placeholder function along those lines, to be replaced when we've made a real one

Reply:
The RFC states that there is a default ACK delay exponent of 3 (e.g. if the transport parameter is not used).

Reply:
I've implemented ack delay exponent retrieval